### PR TITLE
Add more UnwrapFuture overloads

### PR DESCRIPTION
**Pull request by beckerhe**
This adds some more overloads to UnwrapFuture. These overloads handle
`ErrorMessageOr<Future<T>>` cases.

- `Future<ErrorMessageOr<Future<T>>>` unwraps to
  `Future<ErrorMessageOr<T>>`
- `Future<ErrorMessageOr<Future<ErrorMessageOr<T>>>>` unwraps to
  `Future<ErrorMessageOr<T>>`

Ideally these situations shouldn't occur in a good designed API, but
refactorings might make them necessary temporarily.

### CODE REVIEW

File: src/OrbitBase/include/OrbitBase/FutureHelpers.h [hunk 1]
```diff
@@ -73,6 +73,82 @@ template <typename T>
...
+template <typename T>
+[[nodiscard]] inline Future<ErrorMessageOr<T>> UnwrapFuture(
+    const Future<ErrorMessageOr<Future<T>>>& outer_future) {
+  orbit_base::Promise<ErrorMessageOr<T>> promise{};
+  auto future = promise.GetFuture();
+
+  auto outer_continuation =
+      [promise = std::move(promise)](const ErrorMessageOr<Future<T>>& outer_result) mutable {
+        if (outer_result.has_error()) {
+          promise.SetResult(outcome::failure(outer_result.error()));
```
Comment (dimitry-) [inline src/OrbitBase/include/OrbitBase/FutureHelpers.h:85]: do you need outcome::failure here? I think you can drop it.

File: src/OrbitBase/include/OrbitBase/FutureHelpers.h [hunk 2]
```diff
@@ -73,6 +73,82 @@ template <typename T>
...
+template <>
+[[nodiscard]] inline Future<ErrorMessageOr<void>> UnwrapFuture(
+    const Future<ErrorMessageOr<Future<void>>>& outer_future) {
+  orbit_base::Promise<ErrorMessageOr<void>> promise{};
+  auto future = promise.GetFuture();
+
+  auto outer_continuation =
+      [promise = std::move(promise)](const ErrorMessageOr<Future<void>>& outer_result) mutable {
+        if (outer_result.has_error()) {
+          promise.SetResult(outcome::failure(outer_result.error()));
```
Comment (dimitry-) [inline src/OrbitBase/include/OrbitBase/FutureHelpers.h:110]: same here.

File: src/OrbitBase/include/OrbitBase/FutureHelpers.h [hunk 3]
```diff
@@ -73,6 +73,82 @@ template <typename T>
   return promise->GetFuture();
 }
 
+template <typename T>
+[[nodiscard]] inline Future<ErrorMessageOr<T>> UnwrapFuture(
```
Comment (dimitry-) [inline src/OrbitBase/include/OrbitBase/FutureHelpers.h:77]: does inline give us anything here?

File: src/OrbitBase/include/OrbitBase/FutureHelpers.h [hunk 4]
```diff
@@ -73,6 +73,82 @@ template <typename T>
...
+    const Future<ErrorMessageOr<Future<ErrorMessageOr<T>>>>& outer_future) {
+  orbit_base::Promise<ErrorMessageOr<T>> promise{};
+  auto future = promise.GetFuture();
+
+  auto outer_continuation =
+      [promise = std::move(promise)](
+          const ErrorMessageOr<Future<ErrorMessageOr<T>>>& outer_result) mutable {
+        if (outer_result.has_error()) {
+          promise.SetResult(outcome::failure(outer_result.error()));
```
Comment (dimitry-) [inline src/OrbitBase/include/OrbitBase/FutureHelpers.h:136]: and here

File: src/OrbitBase/include/OrbitBase/FutureHelpers.h (hunk 1 above)
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/FutureHelpers.h:85]: Unfortunately not. The implicit constructor of the error case is only available if it's unambiguous.

So for example it won't be available for the type `ErrorMessageOr<ErrorMessage>` or when the `T` in `ErrorMessageOr<T>` has a templated constructor. So to make this universally work, we need outcome::failure.

I have another commit (not yet put out) which will add this to other parts of the code as well.

File: src/OrbitBase/include/OrbitBase/FutureHelpers.h (hunk 3 above)
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/FutureHelpers.h:77]: No, just a copy-and-paste artefact. I will remove it.

**Review by dimitry- — APPROVED** [review]

File: src/OrbitBase/include/OrbitBase/FutureHelpers.h (hunk 3 above)
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/FutureHelpers.h]: Done.

File: src/OrbitBase/include/OrbitBase/FutureHelpers.h (hunk 1 above)
Comment (beckerhe) [reply src/OrbitBase/include/OrbitBase/FutureHelpers.h:85]: I'm going to resolve this.

**Pull request merged**